File: wails/frontend/src/store/__tests__/bookTakeover.test.ts
// Handing a book from one of the writer's machines to another.
//
// The rules under test are the ones that decide whether a manuscript survives
// the handover:
//
//   - A book only opens once the bytes beside us hash to what the other machine
//     recorded. The copy a sync client has not finished replacing is a complete,
//     openable, WRONG book, and size alone cannot tell it apart.
//   - Giving up takes the request back. A request left behind gets granted later
//     by a machine with nobody waiting, and that machine saves, releases and
//     closes a book somebody may be sitting in front of.
//   - Handing over saves first, and refuses if the save is not clean. A 'stale'
//     save means somebody is typing here, which is an answer of no.
//   - Nothing may reach the file after the fingerprint is taken.
//
// Environment: plain node, no jsdom. Only the generated Wails bindings are
// mocked; wailsjs/go/models stays real.

import { describe, it, expect, vi, beforeEach } from 'vitest'
import type { BookData } from '../../types/draftline'

const mocks = vi.hoisted(() => ({
  NewBook: vi.fn(),
  OpenBookDialog: vi.fn(),
  PickBookPath: vi.fn(),
  SaveBook: vi.fn(),
  SaveBookAs: vi.fn(),
  SaveBookSnapshots: vi.fn(),
  OpenRecentProject: vi.fn(),
  OpenBookAsCopy: vi.fn(),
  InspectBookLock: vi.fn(),
  AddRecentProject: vi.fn(),
  IndexBook: vi.fn(),
  MergeEntities: vi.fn(),
  SplitEntity: vi.fn(),
  ImportEPUB: vi.fn(),
  ImportDOCX: vi.fn(),
  ShowInfoDialog: vi.fn(),
  CloseBookFile: vi.fn().mockResolvedValue(undefined),
  AnalyzeBook: vi.fn(),
  RequestBookTakeover: vi.fn(),
  BookTakeoverStatus: vi.fn(),
  WithdrawBookTakeover: vi.fn(),
  GrantBookTakeover: vi.fn(),
  DeclineBookTakeover: vi.fn(),
  PendingBookTakeover: vi.fn(),
  SaveBookAside: vi.fn(),
  LoadSettings: vi.fn(),
  SaveSettings: vi.fn(),
  BrowseForDirectory: vi.fn(),
  GetRecentProjects: vi.fn(),
  RemoveRecentProject: vi.fn(),
  ClearRecentProjects: vi.fn(),
}))

vi.mock('../../../wailsjs/go/main/App', () => mocks)

const PATH = 'C:/books/weather-house.draftline'

function makeBook(): BookData {
  return {
    version: '2.0',
    metadata: { title: 'Test Book', author: 'A. Author', isbn: '', publisher: '', created: '', modified: '', word_count: 2 },
    copyright: '',
    front_matter: [],
    body: [{ id: 'ch-one', title: 'Chapter 1', type: 'chapter', content: '<p>original</p>' }],
    back_matter: [],
    file_path: PATH,
    is_indexed: true,
  }
}

// waiting is an unanswered request whose copy here is NOT yet confirmed to be
// the one the holder last wrote; confirmed is the same request once it is.
const waiting = () => ({ asked: true, held_elsewhere: true, answered: false, granted: false, declined: false, arrived: false, unverifiable: false, can_take_over: false, local_bytes: 0, expected_bytes: 0, message: 'Waiting for STUDIO-DESKTOP to answer…' })
const confirmed = () => ({ ...waiting(), can_take_over: true })
const granted = (over: Record<string, unknown> = {}) => ({ ...waiting(), answered: true, granted: true, held_elsewhere: false, message: 'Waiting for the copy from STUDIO-DESKTOP to sync… 1 MB of 3 MB arrived.', local_bytes: 1, expected_bytes: 3, ...over })

let bookStoreMod: typeof import('../bookStore')
const store = () => bookStoreMod.useBookStore.getState()

async function flush() {
  for (let i = 0; i < 20; i++) await Promise.resolve()
}

beforeEach(async () => {
  vi.resetModules()
  for (const fn of Object.values(mocks)) fn.mockReset()
  mocks.AddRecentProject.mockResolvedValue(undefined)
  mocks.CloseBookFile.mockResolvedValue(undefined)
  vi.useFakeTimers()
  bookStoreMod = await import('../bookStore')
})

describe('asking another device for a book', () => {
  it('waits for the copy to arrive before it opens anything', async () => {
    mocks.RequestBookTakeover.mockResolvedValue(waiting())
    // Granted, then still arriving twice, then proven.
    mocks.BookTakeoverStatus
      .mockResolvedValueOnce(granted())
      .mockResolvedValueOnce(granted({ local_bytes: 2 }))
      .mockResolvedValue(granted({ arrived: true, local_bytes: 3, message: 'STUDIO-DESKTOP handed the book over.' }))
    mocks.OpenRecentProject.mockResolvedValue(makeBook())

    bookStoreMod.useBookStore.setState(s => ({ dialogs: { ...s.dialogs, bookLockWarning: { path: PATH, info: { held: true, stale: false, device: 'STUDIO-DESKTOP', platform: 'windows', app: 'Draftline', last_seen: '', message: '' } } } }))
    const asking = store().askDeviceForBook()

    await flush()
    expect(store().isWaitingForDevice).toBe(true)
    // Nothing is opened on the strength of a grant alone.
    expect(mocks.OpenRecentProject).not.toHaveBeenCalled()
    expect(store().openingLabel).toContain('sync')

    await vi.advanceTimersByTimeAsync(3000)
    await asking

    expect(mocks.OpenRecentProject).toHaveBeenCalledWith(PATH)
    expect(store().book?.metadata.title).toBe('Test Book')
    expect(store().isWaitingForDevice).toBe(false)
    // The receipt is spent once the bytes are proven.
    expect(mocks.WithdrawBookTakeover).toHaveBeenCalledWith(PATH)
  })

  it('never opens a book whose fingerprint could not be checked', async () => {
    mocks.RequestBookTakeover.mockResolvedValue(waiting())
    mocks.BookTakeoverStatus.mockResolvedValue(granted({
      unverifiable: true,
      message: 'STUDIO-DESKTOP handed the book over, but this copy could not be checked against it.',
    }))

    bookStoreMod.useBookStore.setState(s => ({ dialogs: { ...s.dialogs, bookLockWarning: { path: PATH, info: { held: true, stale: false, device: 'STUDIO-DESKTOP', platform: '', app: '', last_seen: '', message: '' } } } }))
    await store().askDeviceForBook()

    expect(mocks.OpenRecentProject).not.toHaveBeenCalled()
    expect(store().book).toBeNull()
  })

  it('takes the request back when the writer stops waiting', async () => {
    mocks.RequestBookTakeover.mockResolvedValue(waiting())
    mocks.BookTakeoverStatus.mockResolvedValue(waiting())

    bookStoreMod.useBookStore.setState(s => ({ dialogs: { ...s.dialogs, bookLockWarning: { path: PATH, info: { held: true, stale: false, device: 'STUDIO-DESKTOP', platform: '', app: '', last_seen: '', message: '' } } } }))
    const asking = store().askDeviceForBook()
    await flush()
    expect(store().isWaitingForDevice).toBe(true)

    store().stopWaitingForDevice()
    await vi.advanceTimersByTimeAsync(2000)
    await asking

    expect(mocks.WithdrawBookTakeover).toHaveBeenCalledWith(PATH)
    expect(mocks.OpenRecentProject).not.toHaveBeenCalled()
    expect(store().isWaitingForDevice).toBe(false)
  })

  it('takes the book when nothing answers and the copy here is confirmed', async () => {
    mocks.RequestBookTakeover.mockResolvedValue(confirmed())
    mocks.BookTakeoverStatus.mockResolvedValue(confirmed())
    mocks.OpenRecentProject.mockResolvedValue(makeBook())

    bookStoreMod.useBookStore.setState(s => ({ dialogs: { ...s.dialogs, bookLockWarning: { path: PATH, info: { held: true, stale: false, device: 'STUDIO-DESKTOP', platform: '', app: '', last_seen: '', message: '' } } } }))
    const asking = store().askDeviceForBook()

    await flush()
    // It says what it is about to do, and when.
    expect(store().openingLabel).toMatch(/Taking it in \d+ seconds?/)
    expect(mocks.OpenRecentProject).not.toHaveBeenCalled()

    await vi.advanceTimersByTimeAsync(31_000)
    await asking

    // Opening claims the book, which is what the other machine stands down on.
    expect(mocks.OpenRecentProject).toHaveBeenCalledWith(PATH)
    expect(mocks.WithdrawBookTakeover).toHaveBeenCalledWith(PATH)
    expect(store().book).not.toBeNull()
  })

  it('will not take a book it cannot confirm, however long it waits', async () => {
    // A machine that never answers is a machine that is RUNNING with nobody in
    // front of it: an unattended one stops heartbeating and goes stale, which
    // raises no question at all. So it was recently writing, and the copy here
    // may be the one from before its last save. The clock running out is not
    // permission on its own.
    mocks.RequestBookTakeover.mockResolvedValue(waiting())
    mocks.BookTakeoverStatus.mockResolvedValue(waiting())

    bookStoreMod.useBookStore.setState(s => ({ dialogs: { ...s.dialogs, bookLockWarning: { path: PATH, info: { held: true, stale: false, device: 'STUDIO-DESKTOP', platform: '', app: '', last_seen: '', message: '' } } } }))
    void store().askDeviceForBook()
    await flush()

    await vi.advanceTimersByTimeAsync(120_000)
    await flush()

    expect(mocks.OpenRecentProject).not.toHaveBeenCalled()
    expect(store().book).toBeNull()
    expect(store().isWaitingForDevice).toBe(true)
    store().stopWaitingForDevice()
    await vi.advanceTimersByTimeAsync(2000)
  })

  it('waits as long as the copy needs once it has been granted', async () => {
    // A grant is an answer, so the thirty seconds does not apply to it: a big
    // book crossing a slow connection takes as long as it takes.
    mocks.RequestBookTakeover.mockResolvedValue(waiting())
    mocks.BookTakeoverStatus.mockResolvedValue(granted())

    bookStoreMod.useBookStore.setState(s => ({ dialogs: { ...s.dialogs, bookLockWarning: { path: PATH, info: { held: true, stale: false, device: 'STUDIO-DESKTOP', platform: '', app: '', last_seen: '', message: '' } } } }))
    void store().askDeviceForBook()
    await flush()

    await vi.advanceTimersByTimeAsync(90_000)
    await flush()

    expect(mocks.OpenRecentProject).not.toHaveBeenCalled()
    expect(store().isWaitingForDevice).toBe(true)
    store().stopWaitingForDevice()
    await vi.advanceTimersByTimeAsync(2000)
  })

  it('puts a refusal on screen, where the status bar is not', async () => {
    mocks.RequestBookTakeover.mockResolvedValue(waiting())
    mocks.BookTakeoverStatus.mockResolvedValue({
      ...waiting(), answered: true, declined: true,
      message: 'Somebody is working on STUDIO-DESKTOP, so it kept the book.',
    })

    bookStoreMod.useBookStore.setState(s => ({ dialogs: { ...s.dialogs, bookLockWarning: { path: PATH, info: { held: true, stale: false, device: 'STUDIO-DESKTOP', platform: '', app: '', last_seen: '', message: '' } } } }))
    await store().askDeviceForBook()

    expect(mocks.OpenRecentProject).not.toHaveBeenCalled()
    expect(store().isWaitingForDevice).toBe(false)
    // The launch screen has no status bar, so a refusal reported there is a
    // refusal nobody sees: the overlay just vanishes.
    const report = store().dialogs.handoverReport
    expect(report?.message).toContain('kept the book')
    expect(report?.offerCopy).toBe(true)
  })

  it('opens a copy from the refusal, which is the way out of it', async () => {
    mocks.RequestBookTakeover.mockResolvedValue(waiting())
    mocks.BookTakeoverStatus.mockResolvedValue({
      ...waiting(), answered: true, declined: true,
      message: 'Somebody is working on STUDIO-DESKTOP, so it kept the book.',
    })
    mocks.OpenBookAsCopy.mockResolvedValue(makeBook())

    bookStoreMod.useBookStore.setState(s => ({ dialogs: { ...s.dialogs, bookLockWarning: { path: PATH, info: { held: true, stale: false, device: 'STUDIO-DESKTOP', platform: '', app: '', last_seen: '', message: '' } } } }))
    await store().askDeviceForBook()
    await store().openCopyAfterHandover()
    await flush()

    expect(mocks.OpenBookAsCopy).toHaveBeenCalledWith(PATH)
    expect(store().dialogs.handoverReport).toBeNull()
  })
})

describe('handing a book to another device', () => {
  it('saves before it hands over, and does not save after', async () => {
    mocks.SaveBook.mockResolvedValue({ success: true, file_path: PATH })
    mocks.GrantBookTakeover.mockResolvedValue({ granted: true, declined: false, withdrawn: false, fingerprinted: true, device: 'LAPTOP' })
    bookStoreMod.useBookStore.setState({ book: makeBook(), isDirty: true })

    await store().grantBookToDevice()
    await flush()

    expect(mocks.SaveBook).toHaveBeenCalledTimes(1)
    expect(mocks.GrantBookTakeover).toHaveBeenCalledTimes(1)
    expect(store().book).toBeNull()
    expect(mocks.CloseBookFile).toHaveBeenCalled()

    // An autosave armed before the handover must not write over the book the
    // other machine has already been told the hash of.
    await vi.advanceTimersByTimeAsync(60_000)
    await flush()
    expect(mocks.SaveBook).toHaveBeenCalledTimes(1)
  })

  it('keeps the book when somebody is typing in it', async () => {
    // A save that completes after another edit comes back stale.
    mocks.SaveBook.mockImplementation(async () => {
      store().updateCurrentContent('<p>still writing</p>')
      return { success: true, file_path: PATH }
    })
    mocks.DeclineBookTakeover.mockResolvedValue({ granted: false, declined: true, withdrawn: false, fingerprinted: false, device: 'LAPTOP' })
    bookStoreMod.useBookStore.setState({ book: makeBook(), isDirty: true })

    await store().grantBookToDevice()
    await flush()

    expect(mocks.GrantBookTakeover).not.toHaveBeenCalled()
    expect(mocks.DeclineBookTakeover).toHaveBeenCalledTimes(1)
    expect(store().book).not.toBeNull()
  })

  it('keeps the book when the other device stopped asking', async () => {
    mocks.GrantBookTakeover.mockResolvedValue({ granted: false, declined: false, withdrawn: true, fingerprinted: false })
    bookStoreMod.useBookStore.setState({ book: makeBook(), isDirty: false })

    await store().grantBookToDevice()
    await flush()

    expect(store().book).not.toBeNull()
    expect(mocks.CloseBookFile).not.toHaveBeenCalled()
  })
})

describe('coming back to a handover already granted', () => {
  it('resumes the wait rather than opening the copy it has', async () => {
    mocks.BookTakeoverStatus
      .mockResolvedValueOnce(granted())
      .mockResolvedValueOnce(granted({ local_bytes: 2 }))
      .mockResolvedValue(granted({ arrived: true, local_bytes: 3 }))
    mocks.OpenRecentProject.mockResolvedValue(makeBook())

    const opening = store().openRecentBook(PATH)
    await flush()
    // The grant beside the book is picked up without asking again.
    expect(mocks.RequestBookTakeover).not.toHaveBeenCalled()
    expect(store().isWaitingForDevice).toBe(true)

    await vi.advanceTimersByTimeAsync(2000)
    await opening
    expect(mocks.OpenRecentProject).toHaveBeenCalledWith(PATH)
  })
})

describe('waking to a book another device has taken', () => {
  const taken = { held: true, stale: false, device: 'STUDIO-DESKTOP', platform: 'windows', app: 'Draftline', last_seen: '', message: 'STUDIO-DESKTOP has this book now, so it was closed here.' }

  it('never writes to the original, and keeps unsaved words in a file of their own', async () => {
    mocks.SaveBookAside.mockResolvedValue({ success: true, file_path: 'C:/books/weather-house (unsaved on LAPTOP).draftline' })
    bookStoreMod.useBookStore.setState({ book: makeBook(), isDirty: true })

    await bookStoreMod.useBookStore.getState().standDownFromBook(taken as never)
    await flush()

    // The original belongs to the other machine now.
    expect(mocks.SaveBook).not.toHaveBeenCalled()
    expect(mocks.SaveBookAside).toHaveBeenCalledTimes(1)
    expect(store().book).toBeNull()
    expect(mocks.CloseBookFile).toHaveBeenCalled()

    // And the autosave that survived the sleep must not fire into the file.
    await vi.advanceTimersByTimeAsync(60_000)
    await flush()
    expect(mocks.SaveBook).not.toHaveBeenCalled()
  })

  it('closes a clean book without leaving a rescue file behind', async () => {
    bookStoreMod.useBookStore.setState({ book: makeBook(), isDirty: false })

    await bookStoreMod.useBookStore.getState().standDownFromBook(taken as never)
    await flush()

    expect(mocks.SaveBookAside).not.toHaveBeenCalled()
    expect(mocks.SaveBook).not.toHaveBeenCalled()
    expect(store().book).toBeNull()
  })

  it('still closes the book when the rescue file cannot be written', async () => {
    mocks.SaveBookAside.mockRejectedValue(new Error('the folder is gone'))
    bookStoreMod.useBookStore.setState({ book: makeBook(), isDirty: true })

    await bookStoreMod.useBookStore.getState().standDownFromBook(taken as never)
    await flush()

    // Staying open would be worse: the next autosave would overwrite the other
    // machine's work.
    expect(store().book).toBeNull()
    expect(mocks.SaveBook).not.toHaveBeenCalled()
  })
})

describe('the other computer closed the book a moment ago', () => {
  // Draftline was shut over there, so its claim is deleted and the deletion is
  // still crossing. Nothing is running to answer, and nothing is running to say
  // what it has — so waiting for either would be waiting for good.
  const letGo = () => ({ ...waiting(), held_elsewhere: false })

  it('opens the book once the claim is gone, without waiting out the clock', async () => {
    mocks.RequestBookTakeover.mockResolvedValue(waiting())
    mocks.BookTakeoverStatus.mockResolvedValue(letGo())
    mocks.OpenRecentProject.mockResolvedValue(makeBook())

    bookStoreMod.useBookStore.setState(s => ({ dialogs: { ...s.dialogs, bookLockWarning: { path: PATH, info: { held: true, stale: false, device: 'STUDIO-DESKTOP', platform: '', app: '', last_seen: '', message: '' } } } }))
    const asking = store().askDeviceForBook()

    // Well inside the thirty seconds, and with nothing ever stamped.
    await vi.advanceTimersByTimeAsync(4000)
    await asking

    expect(mocks.OpenRecentProject).toHaveBeenCalledWith(PATH)
    expect(store().book).not.toBeNull()
    expect(mocks.WithdrawBookTakeover).toHaveBeenCalledWith(PATH)
  })

  it('does not take a book off anybody over one unreadable look', async () => {
    // A sidecar caught mid-write reads as missing. One flicker is not a release.
    mocks.RequestBookTakeover.mockResolvedValue(waiting())
    mocks.BookTakeoverStatus
      .mockResolvedValueOnce(letGo())
      .mockResolvedValue(waiting())

    bookStoreMod.useBookStore.setState(s => ({ dialogs: { ...s.dialogs, bookLockWarning: { path: PATH, info: { held: true, stale: false, device: 'STUDIO-DESKTOP', platform: '', app: '', last_seen: '', message: '' } } } }))
    void store().askDeviceForBook()

    await vi.advanceTimersByTimeAsync(10_000)
    await flush()

    expect(mocks.OpenRecentProject).not.toHaveBeenCalled()
    expect(store().isWaitingForDevice).toBe(true)
    store().stopWaitingForDevice()
    await vi.advanceTimersByTimeAsync(2000)
  })
})

File: wails/booktakeover_test.go
package main

import (
	"context"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"draftline/internal/booklock"
)

// asking is the writer's other machine. It must not be named after this one:
// a claim naming this device is read as our own.
func asking() booklock.Identity {
	return booklock.Identity{Device: "ASKING-MACHINE", Platform: "linux", App: "Draftline 0.21", Session: "s-asking"}
}

func studio() booklock.Identity {
	return booklock.Identity{Device: "STUDIO-DESKTOP", Platform: "windows", App: "Draftline 0.21", Session: "s-studio"}
}

// bookFile is a book with some bytes in it, so there is something to hash.
func bookFile(t *testing.T) string {
	t.Helper()
	path := filepath.Join(t.TempDir(), "The Weather House.draftline")
	if err := os.WriteFile(path, []byte(strings.Repeat("chapter", 400)), 0o644); err != nil {
		t.Fatal(err)
	}
	return path
}

// holdingApp is this machine with the book open and claimed. The claim is taken
// directly rather than through claimDeviceLock, so no heartbeat goroutine is
// left running against a context that is not a Wails one.
func holdingApp(t *testing.T, archive string) *App {
	t.Helper()
	app := &App{ctx: context.Background()}
	lock, _, err := booklock.Force(archive, "bk-1", app.deviceIdentity())
	if err != nil {
		t.Fatalf("claiming the book failed: %v", err)
	}
	app.device.lock = lock
	app.setCurrentFile(archive)
	return app
}

// askFor writes the request the other machine would have written, aimed at
// whoever holds the book.
func askFor(t *testing.T, archive string, id booklock.Identity) {
	t.Helper()
	if _, err := booklock.RequestTakeover(archive, id, booklock.Inspect(archive, id)); err != nil {
		t.Fatalf("RequestTakeover: %v", err)
	}
}

// The handover: the book is saved by then, so granting records what is on disk
// and drops the claim the other machine is watching for.
func TestGrantingHandsTheBookOverWithProof(t *testing.T) {
	archive := bookFile(t)
	app := holdingApp(t, archive)
	askFor(t, archive, asking())

	result := app.GrantBookTakeover()
	if !result.Granted || result.Error != "" {
		t.Fatalf("result = %+v, want a grant", result)
	}
	if !result.Fingerprinted {
		t.Fatal("a readable book must be handed over with a fingerprint")
	}
	if _, err := os.Stat(booklock.SidecarFor(archive)); !os.IsNotExist(err) {
		t.Fatal("the claim should be gone so the other machine can take it")
	}

	reply, ok := booklock.ReadTakeover(archive)
	if !ok || reply.Status != booklock.TakeoverGranted {
		t.Fatalf("reply = %+v, want a grant beside the book", reply)
	}
	if state := booklock.Arrival(archive, reply); !state.Arrived {
		t.Fatalf("the book beside the grant should verify: %+v", state)
	}
}

// The guard that matters most on this side: a writer who gave up takes their
// request with them, and the countdown that was already running must not save,
// release and close a book on a machine with nobody sitting at it.
func TestGrantingWithNobodyWaitingKeepsTheBook(t *testing.T) {
	archive := bookFile(t)
	app := holdingApp(t, archive)
	askFor(t, archive, asking())
	booklock.WithdrawTakeover(archive)

	result := app.GrantBookTakeover()
	if !result.Withdrawn {
		t.Fatalf("result = %+v, want the handover called off", result)
	}
	if result.Granted {
		t.Fatal("nothing should have been handed to nobody")
	}
	if _, err := os.Stat(booklock.SidecarFor(archive)); err != nil {
		t.Fatal("the claim should still be held")
	}
}

// Refusing keeps the book and the claim, and says so where the other machine
// will read it.
func TestDecliningKeepsTheBookAndTheClaim(t *testing.T) {
	archive := bookFile(t)
	app := holdingApp(t, archive)
	askFor(t, archive, asking())

	result := app.DeclineBookTakeover()
	if !result.Declined || result.Error != "" {
		t.Fatalf("result = %+v, want a refusal", result)
	}
	if _, err := os.Stat(booklock.SidecarFor(archive)); err != nil {
		t.Fatal("refusing must not give the claim up")
	}
	reply, ok := booklock.ReadTakeover(archive)
	if !ok || reply.Status != booklock.TakeoverDeclined {
		t.Fatalf("reply = %+v, want a refusal beside the book", reply)
	}
}

// A request aimed at a session that has gone must not be answered by whoever
// holds the book now.
func TestARequestForAnotherSessionIsNotAnswered(t *testing.T) {
	archive := bookFile(t)
	app := holdingApp(t, archive)

	// Aimed at the machine that held it before this session did.
	if _, err := booklock.RequestTakeover(archive, asking(), &booklock.Holder{
		Device: "SOME-OTHER-MACHINE", Session: "s-long-gone",
	}); err != nil {
		t.Fatalf("RequestTakeover: %v", err)
	}
	if result := app.GrantBookTakeover(); !result.Withdrawn {
		t.Fatalf("result = %+v, want it left alone", result)
	}
	if _, err := os.Stat(booklock.SidecarFor(archive)); err != nil {
		t.Fatal("the claim should still be held")
	}
}

// The asking side, mid-sync: the grant has arrived because it is 300 bytes, the
// manuscript has not because it is not. This is what the waiting screen reads,
// and it must not call the book ready.
func TestStatusWaitsWhileTheCopyIsStillArriving(t *testing.T) {
	archive := bookFile(t)
	app := &App{ctx: context.Background()}

	if _, _, err := booklock.Force(archive, "bk-1", studio()); err != nil {
		t.Fatalf("seeding the claim failed: %v", err)
	}
	askFor(t, archive, app.deviceIdentity())
	if _, err := booklock.GrantTakeover(archive, studio()); err != nil {
		t.Fatalf("GrantTakeover: %v", err)
	}

	full, err := os.ReadFile(archive)
	if err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(archive, full[:len(full)/3], 0o644); err != nil {
		t.Fatal(err)
	}

	status := app.BookTakeoverStatus(archive)
	if !status.Granted || !status.Answered {
		t.Fatalf("status = %+v, want an answered grant", status)
	}
	if status.Arrived {
		t.Fatal("a third of a book must not be called arrived")
	}
	if status.ExpectedBytes <= status.LocalBytes {
		t.Fatalf("status = %+v, want it counting up towards the full size", status)
	}
	if !strings.Contains(status.Message, "sync") {
		t.Fatalf("Message = %q, want it to say what it is waiting for", status.Message)
	}

	// And once the bytes land, the same poll says so.
	if err := os.WriteFile(archive, full, 0o644); err != nil {
		t.Fatal(err)
	}
	if status := app.BookTakeoverStatus(archive); !status.Arrived {
		t.Fatalf("status = %+v, want the book ready to open", status)
	}
}

// A refusal has to reach the asking side as a refusal. Silence is a machine
// that is asleep, and the two must not read the same.
func TestStatusReportsARefusal(t *testing.T) {
	archive := bookFile(t)
	app := &App{ctx: context.Background()}

	if _, _, err := booklock.Force(archive, "bk-1", studio()); err != nil {
		t.Fatalf("seeding the claim failed: %v", err)
	}
	askFor(t, archive, app.deviceIdentity())

	if status := app.BookTakeoverStatus(archive); status.Answered {
		t.Fatalf("status = %+v, want it still waiting", status)
	}
	if _, err := booklock.DeclineTakeover(archive, studio()); err != nil {
		t.Fatalf("DeclineTakeover: %v", err)
	}

	status := app.BookTakeoverStatus(archive)
	if !status.Declined || !status.Answered {
		t.Fatalf("status = %+v, want a refusal", status)
	}
	if status.Arrived {
		t.Fatal("a refusal hands nothing over")
	}
}

// Asking for a book nobody has claimed is harmless, and the status says there
// is nothing to wait for.
func TestStatusOfABookNobodyAskedAbout(t *testing.T) {
	archive := bookFile(t)
	app := &App{ctx: context.Background()}
	status := app.BookTakeoverStatus(archive)
	if status.Asked || status.Answered || status.HeldElsewhere {
		t.Fatalf("status = %+v, want nothing going on", status)
	}
}

// The lid-close case: the claim went stale while this machine slept, another
// device took the book, and on waking this session must notice it is no longer
// the holder.
func TestWakingToAStolenBookStandsDown(t *testing.T) {
	archive := bookFile(t)
	app := holdingApp(t, archive)

	// What the machine wakes up to: somebody else's claim on its open book.
	if _, _, err := booklock.Force(archive, "bk-1", studio()); err != nil {
		t.Fatalf("seeding the other claim failed: %v", err)
	}

	if holder := app.claimTakenElsewhere(archive); holder != nil {
		t.Fatalf("one look is not enough to close a book: %+v", holder)
	}
	holder := app.claimTakenElsewhere(archive)
	if holder == nil {
		t.Fatal("two looks agreeing should stand the session down")
	}
	if holder.Device != "STUDIO-DESKTOP" {
		t.Fatalf("Device = %q, want whoever has it now", holder.Device)
	}
	// Standing down must not disturb the claim that is now somebody else's.
	if current := booklock.Inspect(archive, studio()); current == nil || !current.Mine {
		t.Fatalf("claim = %+v, want the new holder's claim untouched", current)
	}
	// And it says so once, not on every tick.
	if again := app.claimTakenElsewhere(archive); again != nil {
		t.Fatal("the writer should be told once")
	}
}

// A sidecar that cannot be read is not evidence of anything. A sync client
// mid-write looks exactly like this, and closing a book over it would lose
// somebody their afternoon.
func TestAMissingClaimDoesNotCloseTheBook(t *testing.T) {
	archive := bookFile(t)
	app := holdingApp(t, archive)
	if err := os.Remove(booklock.SidecarFor(archive)); err != nil {
		t.Fatal(err)
	}
	for look := 0; look < 5; look++ {
		if holder := app.claimTakenElsewhere(archive); holder != nil {
			t.Fatalf("a missing claim must not stand the session down: %+v", holder)
		}
	}
}

// One odd look followed by our own claim again is a hiccup, not a takeover.
func TestASingleOddLookIsForgotten(t *testing.T) {
	archive := bookFile(t)
	app := holdingApp(t, archive)

	if _, _, err := booklock.Force(archive, "bk-1", studio()); err != nil {
		t.Fatal(err)
	}
	if holder := app.claimTakenElsewhere(archive); holder != nil {
		t.Fatal("one look should not be acted on")
	}
	// The claim comes back as ours, as it would when the sync client finishes.
	if _, _, err := booklock.Force(archive, "bk-1", app.deviceIdentity()); err != nil {
		t.Fatal(err)
	}
	if holder := app.claimTakenElsewhere(archive); holder != nil {
		t.Fatalf("our own claim is not a takeover: %+v", holder)
	}
	// And the count started over, so the next odd look is a first look again.
	if _, _, err := booklock.Force(archive, "bk-1", studio()); err != nil {
		t.Fatal(err)
	}
	if holder := app.claimTakenElsewhere(archive); holder != nil {
		t.Fatal("the run of agreeing looks should have restarted")
	}
}

// Ask, be refused, ask again. The second request has to reach the other machine:
// a refusal is an answer to one question, not a standing one.
func TestAskingAgainAfterARefusalIsAnnouncedAgain(t *testing.T) {
	archive := bookFile(t)
	app := holdingApp(t, archive)
	asker := asking()

	askFor(t, archive, asker)
	if first := app.takeoverToAnnounce(archive); first == nil {
		t.Fatal("the first request was not announced")
	}
	if again := app.takeoverToAnnounce(archive); again != nil {
		t.Fatal("the same request should be announced once, not every five seconds")
	}
	if result := app.DeclineBookTakeover(); !result.Declined {
		t.Fatalf("result = %+v, want a refusal", result)
	}

	// The asking machine gives up on the refusal and asks again. Same machine,
	// same running application, same session: only the time has moved.
	booklock.WithdrawTakeover(archive)
	askFor(t, archive, asker)

	second := app.takeoverToAnnounce(archive)
	if second == nil {
		t.Fatal("a second request from the same application must be announced too")
	}
	if second.Device != asker.Device {
		t.Fatalf("Device = %q, want the machine that asked again", second.Device)
	}
}

// Two asks in a row without a refusal in between are still two asks.
func TestAskingTwiceRunningIsAnnouncedTwice(t *testing.T) {
	archive := bookFile(t)
	app := holdingApp(t, archive)

	askFor(t, archive, asking())
	if app.takeoverToAnnounce(archive) == nil {
		t.Fatal("the first request was not announced")
	}
	askFor(t, archive, asking())
	if app.takeoverToAnnounce(archive) == nil {
		t.Fatal("asking again must reach the writer")
	}
}
